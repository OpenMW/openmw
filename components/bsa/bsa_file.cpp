/*
  OpenMW - The completely unofficial reimplementation of Morrowind
  Copyright (C) 2008-2010  Nicolay Korslund
  Email: < korslund@gmail.com >
  WWW: http://openmw.sourceforge.net/

  This file (bsa_file.cpp) is part of the OpenMW package.

  OpenMW is distributed as free software: you can redistribute it
  and/or modify it under the terms of the GNU General Public License
  version 3, as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  version 3 along with this program. If not, see
  http://www.gnu.org/licenses/ .

 */

#include "bsa_file.hpp"

#include <cassert>

#include <boost/filesystem/path.hpp>
#include <boost/filesystem/fstream.hpp>

using namespace std;
using namespace Bsa;


/// Error handling
void BSAFile::fail(const string &msg)
{
    throw std::runtime_error("BSA Error: " + msg + "\nArchive: " + filename);
}

/// Read header information from the input source
void BSAFile::readHeader()
{
    /*
     * The layout of a BSA archive is as follows:
     *
     * - 12 bytes header, contains 3 ints:
     *         id number - equal to 0x100
     *         dirsize - size of the directory block (see below)
     *         numfiles - number of files
     *
     * ---------- start of directory block -----------
     *
     * - 8 bytes*numfiles, each record contains:
     *         fileSize
     *         offset into data buffer (see below)
     *
     * - 4 bytes*numfiles, each record is an offset into the following name buffer
     *
     * - name buffer, indexed by the previous table, each string is
     *   null-terminated. Size is (dirsize - 12*numfiles).
     *
     * ---------- end of directory block -------------
     *
     * - 8*filenum - hash table block, we currently ignore this
     *
     * ----------- start of data buffer --------------
     *
     * - The rest of the archive is file data, indexed by the
     *   offsets in the directory block. The offsets start at 0 at
     *   the beginning of this buffer.
     *
     */
    assert(!isLoaded);

    namespace bfs = boost::filesystem;
    bfs::ifstream input(bfs::path(filename), std::ios_base::binary);

    // Total archive size
    std::streamoff fsize = 0;
    if(input.seekg(0, std::ios_base::end))
    {
        fsize = input.tellg();
        input.seekg(0);
    }

    if(fsize < 12)
        fail("File too small to be a valid BSA archive");

    // Get essential header numbers
    size_t dirsize, filenum;
    {
        // First 12 bytes
        uint32_t head[3];

        input.read(reinterpret_cast<char*>(head), 12);

        if(head[0] != 0x100)
            fail("Unrecognized BSA header");

        // Total number of bytes used in size/offset-table + filename
        // sections.
        dirsize = head[1];

        // Number of files
        filenum = head[2];
    }

    // Each file must take up at least 21 bytes of data in the bsa. So
    // if files*21 overflows the file size then we are guaranteed that
    // the archive is corrupt.
    if((filenum*21 > unsigned(fsize -12)) || (dirsize+8*filenum > unsigned(fsize -12)) )
        fail("Directory information larger than entire archive");

    // Read the offset info into a temporary buffer
    std::vector<uint32_t> offsets(3*filenum);
    input.read(reinterpret_cast<char*>(&offsets[0]), 12*filenum);

    // Read the string table
    stringBuf.resize(dirsize-12*filenum);
    input.read(&stringBuf[0], stringBuf.size());

    // Check our position
    assert(input.tellg() == std::streampos(12+dirsize));

    // Calculate the offset of the data buffer. All file offsets are
    // relative to this. 12 header bytes + directory + hash table
    // (skipped)
    size_t fileDataOffset = 12 + dirsize + 8*filenum;

    // Set up the the FileStruct table
    files.resize(filenum);
    for(size_t i=0;i<filenum;i++)
    {
        FileStruct &fs = files[i];
        fs.fileSize = offsets[i*2];
        fs.offset = offsets[i*2+1] + fileDataOffset;
        fs.name = &stringBuf[offsets[2*filenum+i]];

        if(fs.offset + fs.fileSize > fsize)
            fail("Archive contains offsets outside itself");

        // Add the file name to the lookup
        lookup[fs.name] = i;
    }

    isLoaded = true;
}

/// Get the index of a given file name, or -1 if not found
int BSAFile::getIndex(const char *str) const
{
    Lookup::const_iterator it = lookup.find(str);
    if(it == lookup.end())
        return -1;

    int res = it->second;
    assert(res >= 0 && (size_t)res < files.size());
    return res;
}

/// Open an archive file.
void BSAFile::open(const string &file)
{
    filename = file;
    readHeader();
}

Files::IStreamPtr BSAFile::getFile(const char *file)
{
    assert(file);
    int i = getIndex(file);
    if(i == -1)
        fail("File not found: " + string(file));

    const FileStruct &fs = files[i];

    return Files::openConstrainedFileStream (filename.c_str (), fs.offset, fs.fileSize);
}

Files::IStreamPtr BSAFile::getFile(const FileStruct *file)
{
    return Files::openConstrainedFileStream (filename.c_str (), file->offset, file->fileSize);
}
