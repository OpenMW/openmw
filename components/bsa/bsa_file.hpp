/*
  OpenMW - The completely unofficial reimplementation of Morrowind
  Copyright (C) 2008-2010  Nicolay Korslund
  Email: < korslund@gmail.com >
  WWW: https://openmw.org/

  This file (bsa_file.h) is part of the OpenMW package.

  OpenMW is distributed as free software: you can redistribute it
  and/or modify it under the terms of the GNU General Public License
  version 3, as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  version 3 along with this program. If not, see
  https://www.gnu.org/licenses/ .

 */

#ifndef BSA_BSA_FILE_H
#define BSA_BSA_FILE_H

#include <stdint.h>
#include <string>
#include <vector>
#include <map>

#include <components/misc/stringops.hpp>

#include <components/files/constrainedfilestream.hpp>


namespace Bsa
{

/**
   This class is used to read "Bethesda Archive Files", or BSAs.
 */
class BSAFile
{
public:
    /// Represents one file entry in the archive
    struct FileStruct
    {
        // File size and offset in file. We store the offset from the
        // beginning of the file, not the offset into the data buffer
        // (which is what is stored in the archive.)
        uint32_t fileSize, offset;

        // Zero-terminated file name
        const char *name;
    };
    typedef std::vector<FileStruct> FileList;

private:
    /// Table of files in this archive
    FileList files;

    /// Filename string buffer
    std::vector<char> stringBuf;

    /// True when an archive has been loaded
    bool isLoaded;

    /// Used for error messages
    std::string filename;

    /// Case insensitive string comparison
    struct iltstr
    {
        bool operator()(const char *s1, const char *s2) const
        { return Misc::StringUtils::ciLess(s1, s2); }
    };

    /** A map used for fast file name lookup. The value is the index into
        the files[] vector above. The iltstr ensures that file name
        checks are case insensitive.
    */
    typedef std::map<const char*, int, iltstr> Lookup;
    Lookup lookup;

    /// Error handling
    void fail(const std::string &msg);

    /// Read header information from the input source
    void readHeader();

    /// Get the index of a given file name, or -1 if not found
    /// @note Thread safe.
    int getIndex(const char *str) const;

public:
    /* -----------------------------------
     * BSA management methods
     * -----------------------------------
     */

    BSAFile()
      : isLoaded(false)
    { }

    /// Open an archive file.
    void open(const std::string &file);

    /* -----------------------------------
     * Archive file routines
     * -----------------------------------
     */

    /// Check if a file exists
    bool exists(const char *file) const
    { return getIndex(file) != -1; }

    /** Open a file contained in the archive. Throws an exception if the
        file doesn't exist.
     * @note Thread safe.
    */
    Files::IStreamPtr getFile(const char *file);

    /** Open a file contained in the archive.
     * @note Thread safe.
    */
    Files::IStreamPtr getFile(const FileStruct* file);

    /// Get a list of all files
    /// @note Thread safe.
    const FileList &getList() const
    { return files; }
};

}

#endif
