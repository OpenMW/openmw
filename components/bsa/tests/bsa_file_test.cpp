#include "../bsa_file.hpp"

/*
  Test of the BSAFile class

  This test requires that data/Morrowind.bsa exists in the root
  directory of OpenMW.

 */

#include <iostream>

using namespace std;
using namespace Bsa;

BSAFile bsa;

void find(const char* file)
{
  cout << "Does file '" << file << "' exist?\n  ";
  if(bsa.exists(file))
    {
      cout << "Yes.\n  ";
      cout << bsa.getFile(file)->size() << " bytes\n";
    }
  else cout << "No.\n";
}

int main()
{
  cout << "Reading Morrowind.bsa\n";
  bsa.open("../../data/Morrowind.bsa");

  const BSAFile::FileList &files = bsa.getList();

  cout << "First 10 files in archive:\n";
  for(int i=0; i<10; i++)
    cout << "  " << files[i].name
         << " (" << files[i].fileSize << " bytes @"
         << files[i].offset << ")\n";

  find("meshes\\r\\xnetch_betty.nif");
  find("humdrum");
}
