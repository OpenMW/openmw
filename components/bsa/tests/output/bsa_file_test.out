Reading Morrowind.bsa
First 10 files in archive:
  meshes\m\probe_journeyman_01.nif (6276 bytes @126646052)
  textures\menu_rightbuttonup_top.dds (256 bytes @218530052)
  textures\menu_rightbuttonup_right.dds (256 bytes @218529796)
  textures\menu_rightbuttonup_left.dds (256 bytes @218529540)
  textures\menu_rightbuttondown_top.dds (256 bytes @218528196)
  meshes\b\b_n_redguard_f_skins.nif (41766 bytes @17809778)
  meshes\b\b_n_redguard_m_skins.nif (41950 bytes @18103107)
  meshes\b\b_n_redguard_f_wrist.nif (2355 bytes @17858132)
  meshes\b\b_n_redguard_m_foot.nif (4141 bytes @17862081)
  meshes\b\b_n_redguard_m_knee.nif (2085 bytes @18098101)
Does file 'meshes\r\xnetch_betty.nif' exist?
  Yes.
  53714 bytes
Does file 'humdrum' exist?
  No.
